vehicle/src/Vehicle/Data/Builtin/Standard/Type.hs

```diff
@@ -188,10 +191,30 @@ restrictStandardDeclType declSort (ident, p) typ = do
         RestrictedNetwork -> ValidNetworkType
 
   let expr = BuiltinExpr p (TypeClass tc) [explicit typ]
-  let origin = InstanceTypeRestrictionOrigin $ TypeRestrictionOrigin env (ident, provenanceOf typ) declSort typ
+  let origin = InstanceTypeRestrictionOrigin $ TypeRestrictionOrigin env (ident, provenanceOf typ) (Left declSort) typ
   _ <- createFreshInstanceConstraint False mempty p origin Irrelevant expr
   return typ
 
+restrictStandardRecordAnnotatedAsTensorType ::
+  forall m.
+  (MonadTypeChecker Builtin m) =>
+  DeclProvenance ->
+  [RecordField (Type Builtin)] ->
+  m ()
+restrictStandardRecordAnnotatedAsTensorType (ident, p) fields = case fields of
+  [] -> return ()
+  ((firstFieldName, firstFieldType) : restFields) -> do
+    env <- getFreeEnv
+    let expr = BuiltinExpr p (TypeClass ValidTensorLikeType) [explicit firstFieldType]
+    let firstFieldIdentifier = ((Identifier (modulePath ident) (nameOf firstFieldName)), provenanceOf firstFieldName)
+    let origin = InstanceTypeRestrictionOrigin $ TypeRestrictionOrigin env firstFieldIdentifier (Right RestrictedRecord) firstFieldType
+    _ <- createFreshInstanceConstraint False mempty p origin Irrelevant expr
+
+    -- Add unification constraints to make sure fields are all of the same type.
+    let fieldTypes = map snd restFields
+    _ <- traverse (createFreshUnificationConstraint p mempty (CheckingInstanceType origin) firstFieldType) fieldTypes
```

Comment:
If we created a fresh `origin` object here, we could actually put the two field names we were checking here to improve the error message? We would need to refine `RestrictedRecord` into two different constructors, one for checking the first field is the right type, and one for checking pairs of types are equal?